paparazzi/src/main/java/app/cash/paparazzi/internal/apng/ApngWriter.kt

```diff
@@ -0,0 +1,276 @@
+      sink.writeFDAT(sequenceNumber++, subRect, image)
+    }
+
+    maxWidth = max(maxWidth, image.width)
+    maxHeight = max(maxHeight, image.height)
+    previousFrame = image
+    frameCount++
+  }
+
+  override fun close() {
+    if (frameCount != totalFrameCount) {
+      throw IllegalStateException("Expected $totalFrameCount total frames, actual count $frameCount")
+    }
+
+    sink.writeIEND()
+    fileHandle.resize(fileHandle.position(sink))
+    sink.close()
+
+    if (ihdrWidth == maxWidth && ihdrHeight == maxHeight) {
+      tempFile.copyTo(file, true)
+    } else {
+      correctFirstFrame()
+    }
+
+    fileHandle.close()
+  }
+
+  /**
+   * The IHDR chunk defines the image size from which all animation frames must be
+   * encapsulated. If over the course of the animation we need to render a larger
+   * image than the first frame we need to resize the first frame such that the
+   * entire animation will fit within. Subsequent animation frames are still correct
+   * and don't need to be re-computed.
+   */
+  private fun correctFirstFrame() {
+    val resizedFirstFrame = firstFrame.resize(maxWidth, maxHeight)
+    val unstableSource = fileHandle.source(iDATChunkEndOffset)
+    val stableHandle = FileSystem.SYSTEM.openReadWrite(file.path.toPath())
+    val stableSink = stableHandle.sink().buffer()
+    stableSink.writePngSignature()
+    stableSink.writeIHDR(resizedFirstFrame)
+    stableSink.writeACTL(frameCount, 0)
+    stableSink.writeFCTL(0, Rectangle(maxWidth, maxHeight))
+    stableSink.writeIDAT(resizedFirstFrame)
+    stableSink.write(unstableSource, fileHandle.size() - iDATChunkEndOffset)
+    stableHandle.resize(stableHandle.position(stableSink))
+    stableSink.close()
+    stableHandle.close()
+    unstableSource.close()
+  }
+
+  private fun BufferedSink.writePngSignature() {
+    write(PngConstants.PNG_SIG)
+  }
+
+  private fun BufferedSink.writeIHDR(bufferedImage: BufferedImage) {
+    val data = Buffer().apply {
+      writeInt(bufferedImage.width)
+      writeInt(bufferedImage.height)
+      ihdrWidth = bufferedImage.width
+      ihdrHeight = bufferedImage.height
+
+      // https://www.w3.org/TR/png/#3colourType
+      val colorType = when (bufferedImage.type) {
+        BufferedImage.TYPE_INT_RGB -> PNG_COLOR_TYPE_RGB
+        BufferedImage.TYPE_INT_ARGB -> PNG_COLOR_TYPE_RGBA
+        else -> throw IllegalStateException("Unsupported image type")
+      }
+
+      writeByte(PNG_BITS_PER_PIXEL.toInt())
+      writeByte(colorType.toInt())
+      writeByte(0) // Compression
+      writeByte(0) // Filter
+      writeByte(0) // Interlace
+    }
+
+    writeChunk(IHDR, data)
+  }
+
+  private fun BufferedSink.writeACTL(frameCount: Int, loopCount: Int) {
+    val data = Buffer().apply {
+      writeInt(frameCount)
+      writeInt(loopCount)
+    }
+
+    writeChunk(ACTL, data)
+  }
+
+  private fun BufferedSink.writeFCTL(
+    sequenceNumber: Int,
+    rectangle: Rectangle
+  ) {
+    val data = Buffer().apply {
```

Comment:
Creating all of these temporary buffers shouldn’t be necessary? What if we changed the signature of writeChunk to this:

```
  private fun BufferedSink.writeChunk(header: Header, content: BufferedSink.() -> {}) {
```

Then you could rewind to save some copies.

For now as-is is okay!